baseplate/thrift.go

```diff
@@ -0,0 +1,153 @@
+package baseplate
+
+import (
+	"context"
+	"io"
+	"time"
+
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/getsentry/raven-go"
+	"github.com/reddit/baseplate.go/edgecontext"
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/secrets"
+	"github.com/reddit/baseplate.go/thriftbp"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+type baseplateThriftServer struct {
+	thriftServer *thrift.TSimpleServer
+	config       ServerConfig
+	afterStop    []io.Closer
+	logger       log.Wrapper
+}
+
+func (bts *baseplateThriftServer) Config() ServerConfig {
+	return bts.config
+}
+
+func (bts *baseplateThriftServer) Impl() interface{} {
+	return bts.thriftServer
+}
+
+func (bts *baseplateThriftServer) Serve() error {
+	return bts.thriftServer.Serve()
+}
+
+func (bts *baseplateThriftServer) Stop() error {
+	var err error
+	bts.thriftServer.Stop()
+	for _, c := range bts.afterStop {
+		c.Close()
+	}
+	return err
+}
+func initLogger(cfg ServerConfig) log.Wrapper {
+	if cfg.Log.Level == "" {
+		cfg.Log.Level = log.InfoLevel
+	}
+	level := cfg.Log.Level
+	log.InitLogger(level)
+	return log.ZapWrapper(level)
+}
+
+func initSecrets(ctx context.Context, cfg ServerConfig, logger log.Wrapper) (*secrets.Store, error) {
+	ctx, cancel := context.WithTimeout(ctx, time.Second*30)
+	defer cancel()
+
+	secretsStore, err := secrets.NewStore(ctx, cfg.Secrets.Path, logger)
+	if err != nil {
+		return nil, err
+	}
+	return secretsStore, nil
+}
+
+func initTracing(cfg ServerConfig, logger log.Wrapper, metrics *metricsbp.Statsd) error {
+	if err := tracing.InitGlobalTracer(tracing.TracerConfig{
+		ServiceName:      cfg.Tracing.Namespace,
+		SampleRate:       cfg.Tracing.SampleRate,
+		Logger:           logger,
+		MaxRecordTimeout: cfg.Tracing.RecordTimeout,
+		QueueName:        "main",
+	}); err != nil {
+		return err
+	}
+
+	tracing.RegisterCreateServerSpanHooks(
+		metricsbp.CreateServerSpanHook{Metrics: metrics},
+		tracing.ErrorReporterCreateServerSpanHook{},
+	)
+	return nil
+}
+
+func initSentry(cfg ServerConfig) error {
+	if err := raven.SetDSN(cfg.Sentry.DSN); err != nil {
+		return err
+	}
+	if err := raven.SetSampleRate(float32(cfg.Sentry.SampleRate)); err != nil {
+		return err
+	}
+	raven.SetEnvironment(cfg.Sentry.Environment)
+	return nil
+}
+
+func cleanup(closers []io.Closer) {
+	for _, c := range closers {
+		if err := c.Close(); err != nil {
+			log.Error("error cleaning up thrift server initialization:", err)
+		}
+	}
+}
+
+// NewBaseplateThriftServer returns a server that includes the default middleware.
+func NewBaseplateThriftServer(ctx context.Context, cfg ServerConfig, processor thriftbp.BaseplateProcessor, additionalMiddlewares ...thriftbp.Middleware) (Server, error) {
```

Comment:
Since the package is already named `baseplate`, this could be just called `NewThriftServer`.

Reply:
Yeah, I'm a little reluctant b/c there is already a NewThriftServer, and I'd not like them confused.  NewConfiguredThriftServer?

Reply:
From user's perspective, it's `baseplate.NewThriftServer` vs. `thriftbp.NewThriftServer`, so it's not very confusing? Also that reminds me the other one probably should be renamed into `thriftbp.NewServer` :)

Reply:
🔕 I do agree with the general principle, and I'd rather type an extra word in this particular case, because it's the user's *first* thing they're doing in the package, and I'd rather not have noobs get confused when the wrong one gets autocompleted.

Reply:
Feel free to rename both. This is something `golint` would complain about (when you have `package.PackageSomething` defined), see also https://github.com/reddit/baseplate.go/pull/137.